src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -31,6 +35,12 @@ internal class InternalEvaluatorImpl(
 
     private val registeredCompiledScripts = arrayListOf<SerializedCompiledScript>()
 
+    /**
+     * Stores cache for val values.
+     * Its contents is a part of variablesHolder.
+     */
+    private val variablesCache = mutableMapOf<String, VariableState>()
```

Comment:
Comment made the things clearer. So, it may be called `valVariablesCache`. Or it also may be a set of names, right? I don't see the reason to store same references in several maps

Reply:
Yes, seems like we might go for reduction here